### ISSUE TITLE
Various crashes while connection is offline or partially connected

### ISSUE THREAD

**Issue by syldrathecat** [fchat-pidgin/fchat-pidgin#128]
Various crashes possible while the account is not connected or partially connecting.

A large number of crashes are caused by not checking the result of `purple_conversation_get_gc()` which may return `NULL`. This causes almost every /command to crash when typed in an offline account's conversation window.

Another large family of crashes are caused by `fla->proper_character` being `NULL` during the partially-connected phase.

Notable crashes appear when typing any /command in to an offline or partially-connected conversation, mousing over links in an offline conversation, or when attempting to open an IM window while partially-connected.

---

To test all possible instances, all commands/features should be attempted while offline, and through different phases of "partially-connected" (before SSL connection established, before ticket request/reply, before IDN request/reply, before VAR messages, before HLO/CON (probably not neccessary), before friends list (FRL), before ignore list (IGN), before admin list (ADL), before all character info (LIS) is received, and before the initial friends/bookmarks sync), preferably by blocking or delaying processing of certain messages to leave the client in partially-connected states for extended periods.

Main features to test are: /commands in conversations (chat and IM), opening/closing IM windows, opening(?)/closing chat windows, requesting character info, changing status, creating a private channel, operating the ignore list, using the "Make Friend Request" or "Add Bookmark" options on a buddy (or their removal counterparts), clicking the alert staff / character info buttons on a conversation.

**Comment by vobe** [comment]
Can we consider this fixed following 699f3d9 and ae1b998 ?

**Comment by syldrathecat** [comment]
Wow thanks! I did some testing with the patches and couldn't get anything obvious to trigger a crash.

I did get a few login failures of the form: "This command requires that you have logged in." by doing stuff during login, but that's a much less painful problem. Funny enough Pidgin blocks typing in messages while connecting, but doesn't block "/me" messages.

A read over the code seems to point to some of the other potential crash issues existing, but it seems to only be in cases where the server sends an unexpected sequence of packets compared to what it does now (e.g. COP must come before ICH, and ADL must come before any COP/JCH messages, due to assumptions that the related pointers aren't null in flist_get_permissions).

I'm pretty confident in saying the issue is resolved now.
